i586/porteus/base/003-mate/usr/man/man5/smb.conf.5
'\" t
.\"     Title: smb.conf
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.78.1 <http://docbook.sf.net/>
.\"      Date: 10/20/2013
.\"    Manual: File Formats and Conventions
.\"    Source: Samba 4.0
.\"  Language: English
.\"
.TH "SMB\&.CONF" "5" "10/20/2013" "Samba 4\&.0" "File Formats and Conventions"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
smb.conf \- The configuration file for the Samba suite
.SH "SYNOPSIS"
.PP
The
smb\&.conf
file is a configuration file for the Samba suite\&.
smb\&.conf
contains runtime configuration information for the Samba programs\&. The complete description of the file format and possible parameters held within are here for reference purposes\&.
.SH "FILE FORMAT"
.PP
The file consists of sections and parameters\&. A section begins with the name of the section in square brackets and continues until the next section begins\&. Sections contain parameters of the form:
.sp
.if n \{\
.RS 4
.\}
.nf
\fIname\fR = \fIvalue \fR
.fi
.if n \{\
.RE
.\}
.PP
The file is line\-based \- that is, each newline\-terminated line represents either a comment, a section name or a parameter\&.
.PP
Section and parameter names are not case sensitive\&.
.PP
Only the first equals sign in a parameter is significant\&. Whitespace before or after the first equals sign is discarded\&. Leading, trailing and internal whitespace in section and parameter names is irrelevant\&. Leading and trailing whitespace in a parameter value is discarded\&. Internal whitespace within a parameter value is retained verbatim\&.
.PP
Any line beginning with a semicolon (\(lq;\(rq) or a hash (\(lq#\(rq) character is ignored, as are lines containing only whitespace\&.
.PP
Any line ending in a
\(lq\e\(rq
is continued on the next line in the customary UNIX fashion\&.
.PP
The values following the equals sign in parameters are all either a string (no quotes needed) or a boolean, which may be given as yes/no, 1/0 or true/false\&. Case is not significant in boolean values, but is preserved in string values\&. Some items such as create masks are numeric\&.
.SH "SECTION DESCRIPTIONS"
.PP
Each section in the configuration file (except for the [global] section) describes a shared resource (known as a
\(lqshare\(rq)\&. The section name is the name of the shared resource and the parameters within the section define the shares attributes\&.
.PP
There are three special sections, [global], [homes] and [printers], which are described under
\fIspecial sections\fR\&. The following notes apply to ordinary section descriptions\&.
.PP
A share consists of a directory to which access is being given plus a description of the access rights which are granted to the user of the service\&. Some housekeeping options are also specifiable\&.
.PP
Sections are either file share services (used by the client as an extension of their native file systems) or printable services (used by the client to access print services on the host running the server)\&.
.PP
Sections may be designated
\fIguest\fR
services, in which case no password is required to access them\&. A specified UNIX
\fIguest account\fR
is used to define access privileges in this case\&.
.PP
Sections other than guest services will require a password to access them\&. The client provides the username\&. As older clients only provide passwords and not usernames, you may specify a list of usernames to check against the password using the
user =
option in the share definition\&. For modern clients such as Windows 95/98/ME/NT/2000, this should not be necessary\&.
.PP
The access rights granted by the server are masked by the access rights granted to the specified or guest UNIX user by the host system\&. The server does not grant more access than the host system grants\&.
.PP
The following sample section defines a file space share\&. The user has write access to the path
/home/bar\&. The share is accessed via the share name
foo:
.sp
.if n \{\
.RS 4
.\}
.nf
	\fI[foo]\fR
	\m[blue]\fBpath = /home/bar\fR\m[]
	\m[blue]\fBread only = no\fR\m[]
.fi
.if n \{\
.RE
.\}
.PP
The following sample section defines a printable share\&. The share is read\-only, but printable\&. That is, the only write access permitted is via calls to open, write to and close a spool file\&. The
\fIguest ok\fR
parameter means access will be permitted as the default guest user (specified elsewhere):
.sp
.if n \{\
.RS 4
.\}
.nf
	\fI[aprinter]\fR
	\m[blue]\fBpath = /usr/spool/public\fR\m[]
	\m[blue]\fBread only = yes\fR\m[]
	\m[blue]\fBprintable = yes\fR\m[]
	\m[blue]\fBguest ok = yes\fR\m[]
.fi
.if n \{\
.RE
.\}
.sp
.SH "SPECIAL SECTIONS"
.SS "The [global] section"
.PP
Parameters in this section apply to the server as a whole, or are defaults for sections that do not specifically define certain items\&. See the notes under PARAMETERS for more information\&.
.SS "The [homes] section"
.PP
If a section called [homes] is included in the configuration file, services connecting clients to their home directories can be created on the fly by the server\&.
.PP
When the connection request is made, the existing sections are scanned\&. If a match is found, it is used\&. If no match is found, the requested section name is treated as a username and looked up in the local password file\&. If the name exists and the correct password has been given, a share is created by cloning the [homes] section\&.
.PP
Some modifications are then made to the newly created share:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
The share name is changed from homes to the located username\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
If no path was given, the path is set to the user\*(Aqs home directory\&.
.RE
.sp
.RE
.PP
If you decide to use a
\fIpath =\fR
line in your [homes] section, it may be useful to use the %S macro\&. For example:
.sp
.if n \{\
.RS 4
.\}
.nf
\fBpath = /data/pchome/%S\fR
.fi
.if n \{\
.RE
.\}
.sp
is useful if you have different home directories for your PCs than for UNIX access\&.
.PP
This is a fast and simple way to give a large number of clients access to their home directories with a minimum of fuss\&.
.PP
A similar process occurs if the requested section name is
\(lqhomes\(rq, except that the share name is not changed to that of the requesting user\&. This method of using the [homes] section works well if different users share a client PC\&.
.PP
The [homes] section can specify all the parameters a normal service section can specify, though some make more sense than others\&. The following is a typical and suitable [homes] section:
.sp
.if n \{\
.RS 4
.\}
.nf
\fI[homes]\fR
\m[blue]\fBread only = no\fR\m[]
.fi
.if n \{\
.RE
.\}
.PP
An important point is that if guest access is specified in the [homes] section, all home directories will be visible to all clients
\fIwithout a password\fR\&. In the very unlikely event that this is actually desirable, it is wise to also specify
\fIread only access\fR\&.
.PP
The
\fIbrowseable\fR
flag for auto home directories will be inherited from the global browseable flag, not the [homes] browseable flag\&. This is useful as it means setting
\fIbrowseable = no\fR
in the [homes] section will hide the [homes] share but make any auto home directories visible\&.
.SS "The [printers] section"
.PP
This section works like [homes], but for printers\&.
.PP
If a [printers] section occurs in the configuration file, users are able to connect to any printer specified in the local host\*(Aqs printcap file\&.
.PP
When a connection request is made, the existing sections are scanned\&. If a match is found, it is used\&. If no match is found, but a [homes] section exists, it is used as described above\&. Otherwise, the requested section name is treated as a printer name and the appropriate printcap file is scanned to see if the requested section name is a valid printer share name\&. If a match is found, a new printer share is created by cloning the [printers] section\&.
.PP
A few modifications are then made to the newly created share:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
The share name is set to the located printer name
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
If no printer name was given, the printer name is set to the located printer name
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
If the share does not permit guest access and no username was given, the username is set to the located printer name\&.
.RE
.sp
.RE
.PP
The [printers] service MUST be printable \- if you specify otherwise, the server will refuse to load the configuration file\&.
.PP
Typically the path specified is that of a world\-writeable spool directory with the sticky bit set on it\&. A typical [printers] entry looks like this:
.sp
.if n \{\
.RS 4
.\}
.nf
\fI[printers]\fR
\m[blue]\fBpath = /usr/spool/public\fR\m[]
\m[blue]\fBguest ok = yes\fR\m[]
\m[blue]\fBprintable = yes\fR\m[]
.fi
.if n \{\
.RE
.\}
.PP
All aliases given for a printer in the printcap file are legitimate printer names as far as the server is concerned\&. If your printing subsystem doesn\*(Aqt work like that, you will have to set up a pseudo\-printcap\&. This is a file consisting of one or more lines like this:
.sp
.if n \{\
.RS 4
.\}
.nf
alias|alias|alias|alias\&.\&.\&.    
.fi
.if n \{\
.RE
.\}
.PP
Each alias should be an acceptable printer name for your printing subsystem\&. In the [global] section, specify the new file as your printcap\&. The server will only recognize names found in your pseudo\-printcap, which of course can contain whatever aliases you like\&. The same technique could be used simply to limit access to a subset of your local printers\&.
.PP
An alias, by the way, is defined as any component of the first entry of a printcap record\&. Records are separated by newlines, components (if there are more than one) are separated by vertical bar symbols (|)\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.PP
On SYSV systems which use lpstat to determine what printers are defined on the system you may be able to use
printcap name = lpstat
to automatically obtain a list of printers\&. See the
printcap name
option for more details\&.
.sp .5v
.RE
.SH "USERSHARES"
.PP
Starting with Samba version 3\&.0\&.23 the capability for non\-root users to add, modify, and delete their own share definitions has been added\&. This capability is called
\fIusershares\fR
and is controlled by a set of parameters in the [global] section of the smb\&.conf\&. The relevant parameters are :
.PP
usershare allow guests
.RS 4
Controls if usershares can permit guest access\&.
.RE
.PP
usershare max shares
.RS 4
Maximum number of user defined shares allowed\&.
.RE
.PP
usershare owner only
.RS 4
If set only directories owned by the sharing user can be shared\&.
.RE
.PP
usershare path
.RS 4
Points to the directory containing the user defined share definitions\&. The filesystem permissions on this directory control who can create user defined shares\&.
.RE
.PP
usershare prefix allow list
.RS 4
Comma\-separated list of absolute pathnames restricting what directories can be shared\&. Only directories below the pathnames in this list are permitted\&.
.RE
.PP
usershare prefix deny list
.RS 4
Comma\-separated list of absolute pathnames restricting what directories can be shared\&. Directories below the pathnames in this list are prohibited\&.
.RE
.PP
usershare template share
.RS 4
Names a pre\-existing share used as a template for creating new usershares\&. All other share parameters not specified in the user defined share definition are copied from this named share\&.
.RE
.PP
To allow members of the UNIX group
foo
to create user defined shares, create the directory to contain the share definitions as follows:
.PP
Become root:
.sp
.if n \{\
.RS 4
.\}
.nf
mkdir /usr/local/samba/lib/usershares
chgrp foo /usr/local/samba/lib/usershares
chmod 1770 /usr/local/samba/lib/usershares
.fi
.if n \{\
.RE
.\}
.PP
Then add the parameters
.sp
.if n \{\
.RS 4
.\}
.nf
	\m[blue]\fBusershare path = /usr/local/samba/lib/usershares\fR\m[]
	\m[blue]\fBusershare max shares = 10\fR\m[] # (or the desired number of shares)
.fi
.if n \{\
.RE
.\}
.sp
to the global section of your
smb\&.conf\&. Members of the group foo may then manipulate the user defined shares using the following commands\&.
.PP
net usershare add sharename path [comment] [acl] [guest_ok=[y|n]]
.RS 4
To create or modify (overwrite) a user defined share\&.
.RE
.PP
net usershare delete sharename
.RS 4
To delete a user defined share\&.
.RE
.PP
net usershare list wildcard\-sharename
.RS 4
To list user defined shares\&.
.RE
.PP
net usershare info wildcard\-sharename
.RS 4
To print information about user defined shares\&.
.RE
.SH "PARAMETERS"
.PP
Parameters define the specific attributes of sections\&.
.PP
Some parameters are specific to the [global] section (e\&.g\&.,
\fIsecurity\fR)\&. Some parameters are usable in all sections (e\&.g\&.,
\fIcreate mask\fR)\&. All others are permissible only in normal sections\&. For the purposes of the following descriptions the [homes] and [printers] sections will be considered normal\&. The letter
\fIG\fR
in parentheses indicates that a parameter is specific to the [global] section\&. The letter
\fIS\fR
indicates that a parameter can be specified in a service specific section\&. All
\fIS\fR
parameters can also be specified in the [global] section \- in which case they will define the default behavior for all services\&.
.PP
Parameters are arranged here in alphabetical order \- this may not create best bedfellows, but at least you can find them! Where there are synonyms, the preferred synonym is described, others refer to the preferred synonym\&.
.SH "VARIABLE SUBSTITUTIONS"
.PP
Many of the strings that are settable in the config file can take substitutions\&. For example the option
\(lqpath = /tmp/%u\(rq
is interpreted as
\(lqpath = /tmp/john\(rq
if the user connected with the username john\&.
.PP
These substitutions are mostly noted in the descriptions below, but there are some general substitutions which apply whenever they might be relevant\&. These are:
.PP
%U
.RS 4
session username (the username that the client wanted, not necessarily the same as the one they got)\&.
.RE
.PP
%G
.RS 4
primary group name of %U\&.
.RE
.PP
%h
.RS 4
the Internet hostname that Samba is running on\&.
.RE
.PP
%m
.RS 4
the NetBIOS name of the client machine (very useful)\&.
.sp
This parameter is not available when Samba listens on port 445, as clients no longer send this information\&. If you use this macro in an include statement on a domain that has a Samba domain controller be sure to set in the [global] section
\fIsmb ports = 139\fR\&. This will cause Samba to not listen on port 445 and will permit include functionality to function as it did with Samba 2\&.x\&.
.RE
.PP
%L
.RS 4
the NetBIOS name of the server\&. This allows you to change your config based on what the client calls you\&. Your server can have a
\(lqdual personality\(rq\&.
.RE
.PP
%M
.RS 4
the Internet name of the client machine\&.
.RE
.PP
%R
.RS 4
the selected protocol level after protocol negotiation\&. It can be one of CORE, COREPLUS, LANMAN1, LANMAN2, NT1, SMB2_02, SMB2_10, SMB2_22, SMB2_24, SMB3_00 or SMB2_FF\&.
.RE
.PP
%d
.RS 4
the process id of the current server process\&.
.RE
.PP
%a
.RS 4
The architecture of the remote machine\&. It currently recognizes Samba (\fBSamba\fR), the Linux CIFS file system (\fBCIFSFS\fR), OS/2, (\fBOS2\fR), Mac OS X (\fBOSX\fR), Windows for Workgroups (\fBWfWg\fR), Windows 9x/ME (\fBWin95\fR), Windows NT (\fBWinNT\fR), Windows 2000 (\fBWin2K\fR), Windows XP (\fBWinXP\fR), Windows XP 64\-bit(\fBWinXP64\fR), Windows 2003 including 2003R2 (\fBWin2K3\fR), and Windows Vista (\fBVista\fR)\&. Anything else will be known as
\fBUNKNOWN\fR\&.
.RE
.PP
%I
.RS 4
the IP address of the client machine\&.
.sp
Before 4\&.0\&.0 it could contain IPv4 mapped IPv6 addresses, now it only contains IPv4 or IPv6 addresses\&.
.RE
.PP
%i
.RS 4
the local IP address to which a client connected\&.
.sp
Before 4\&.0\&.0 it could contain IPv4 mapped IPv6 addresses, now it only contains IPv4 or IPv6 addresses\&.
.RE
.PP
%T
.RS 4
the current date and time\&.
.RE
.PP
%D
.RS 4
name of the domain or workgroup of the current user\&.
.RE
.PP
%w
.RS 4
the winbind separator\&.
.RE
.PP
%$(\fIenvvar\fR)
.RS 4
the value of the environment variable
\fIenvar\fR\&.
.RE
.PP
The following substitutes apply only to some configuration options (only those that are used when a connection has been established):
.PP
%S
.RS 4
the name of the current service, if any\&.
.RE
.PP
%P
.RS 4
the root directory of the current service, if any\&.
.RE
.PP
%u
.RS 4
username of the current service, if any\&.
.RE
.PP
%g
.RS 4
primary group name of %u\&.
.RE
.PP
%H
.RS 4
the home directory of the user given by %u\&.
.RE
.PP
%N
.RS 4
the name of your NIS home directory server\&. This is obtained from your NIS auto\&.map entry\&. If you have not compiled Samba with the
\fI\-\-with\-automount\fR
option, this value will be the same as %L\&.
.RE
.PP
%p
.RS 4
the path of the service\*(Aqs home directory, obtained from your NIS auto\&.map entry\&. The NIS auto\&.map entry is split up as
%N:%p\&.
.RE
.PP
There are some quite creative things that can be done with these substitutions and other
smb\&.conf
options\&.
.SH "NAME MANGLING"
.PP
Samba supports
name mangling
so that DOS and Windows clients can use files that don\*(Aqt conform to the 8\&.3 format\&. It can also be set to adjust the case of 8\&.3 format filenames\&.
.PP
There are several options that control the way mangling is performed, and they are grouped here rather than listed separately\&. For the defaults look at the output of the testparm program\&.
.PP
These options can be set separately for each service\&.
.PP
The options are:
.PP
case sensitive = yes/no/auto
.RS 4
controls whether filenames are case sensitive\&. If they aren\*(Aqt, Samba must do a filename search and match on passed names\&. The default setting of auto allows clients that support case sensitive filenames (Linux CIFSVFS and smbclient 3\&.0\&.5 and above currently) to tell the Samba server on a per\-packet basis that they wish to access the file system in a case\-sensitive manner (to support UNIX case sensitive semantics)\&. No Windows or DOS system supports case\-sensitive filename so setting this option to auto is that same as setting it to no for them\&. Default
\fIauto\fR\&.
.RE
.PP
default case = upper/lower
.RS 4
controls what the default case is for new filenames (ie\&. files that don\*(Aqt currently exist in the filesystem)\&. Default
\fIlower\fR\&. IMPORTANT NOTE: As part of the optimizations for directories containing large numbers of files, the following special case applies\&. If the options
\m[blue]\fBcase sensitive = yes\fR\m[],
\m[blue]\fBpreserve case = No\fR\m[], and
\m[blue]\fBshort preserve case = No\fR\m[]
are set, then the case of
\fIall\fR
incoming client filenames, not just new filenames, will be modified\&. See additional notes below\&.
.RE
.PP
preserve case = yes/no
.RS 4
controls whether new files (ie\&. files that don\*(Aqt currently exist in the filesystem) are created with the case that the client passes, or if they are forced to be the
default
case\&. Default
\fIyes\fR\&.
.RE
.PP
short preserve case = yes/no
.RS 4
controls if new files (ie\&. files that don\*(Aqt currently exist in the filesystem) which conform to 8\&.3 syntax, that is all in upper case and of suitable length, are created upper case, or if they are forced to be the
default
case\&. This option can be used with
preserve case = yes
to permit long filenames to retain their case, while short names are lowercased\&. Default
\fIyes\fR\&.
.RE
.PP
By default, Samba 3\&.0 has the same semantics as a Windows NT server, in that it is case insensitive but case preserving\&. As a special case for directories with large numbers of files, if the case options are set as follows, "case sensitive = yes", "case preserve = no", "short preserve case = no" then the "default case" option will be applied and will modify all filenames sent from the client when accessing this share\&.
.SH "REGISTRY-BASED CONFIGURATION"
.PP
Starting with Samba version 3\&.2\&.0, the capability to store Samba configuration in the registry is available\&. The configuration is stored in the registry key
\fIHKLM\eSoftware\eSamba\esmbconf\fR\&. There are two levels of registry configuration:
.sp
.RS 4
.ie n \{\
\h'-04' 1.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  1." 4.2
.\}
Share definitions stored in registry are used\&. This is triggered by setting the global parameter
\fIregistry shares\fR
to
\(lqyes\(rq
in
\fIsmb\&.conf\fR\&.
.sp
The registry shares are loaded not at startup but on demand at runtime by
\fIsmbd\fR\&. Shares defined in
\fIsmb\&.conf\fR
take priority over shares of the same name defined in registry\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 2.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  2." 4.2
.\}
Global
\fIsmb\&.conf\fR
options stored in registry are used\&. This can be activated in two different ways:
.sp
Firstly, a registry only configuration is triggered by setting
\m[blue]\fBconfig backend = registry\fR\m[]
in the [global] section of
\fIsmb\&.conf\fR\&. This resets everything that has been read from config files to this point and reads the content of the global configuration section from the registry\&. This is the recommended method of using registry based configuration\&.
.sp
Secondly, a mixed configuration can be activated by a special new meaning of the parameter
\m[blue]\fBinclude = registry\fR\m[]
in the [global] section of
\fIsmb\&.conf\fR\&. This reads the global options from registry with the same priorities as for an include of a text file\&. This may be especially useful in cases where an initial configuration is needed to access the registry\&.
.sp
Activation of global registry options automatically activates registry shares\&. So in the registry only case, shares are loaded on demand only\&.
.RE
.sp
.RE
.PP
Note: To make registry\-based configurations foolproof at least to a certain extent, the use of
\fIlock directory\fR
and
\fIconfig backend\fR
inside the registry configuration has been disabled: Especially by changing the
\fIlock directory\fR
inside the registry configuration, one would create a broken setup where the daemons do not see the configuration they loaded once it is active\&.
.PP
The registry configuration can be accessed with tools like
\fIregedit\fR
or
\fInet (rpc) registry\fR
in the key
\fIHKLM\eSoftware\eSamba\esmbconf\fR\&. More conveniently, the
\fIconf\fR
subcommand of the
\fBnet\fR(8)
utility offers a dedicated interface to read and write the registry based configuration locally, i\&.e\&. directly accessing the database file, circumventing the server\&.
.SH "EXPLANATION OF EACH PARAMETER"


bind interfaces only (G)
.\" bind interfaces only
.PP
.RS 4
This global parameter allows the Samba admin to limit what interfaces on a machine will serve SMB requests\&. It affects file service
\fBsmbd\fR(8)
and name service
\fBnmbd\fR(8)
in a slightly different ways\&.
.sp
For name service it causes
nmbd
to bind to ports 137 and 138 on the interfaces listed in the
\m[blue]\fBinterfaces\fR\m[]
parameter\&.
nmbd
also binds to the "all addresses" interface (0\&.0\&.0\&.0) on ports 137 and 138 for the purposes of reading broadcast messages\&. If this option is not set then
nmbd
will service name requests on all of these sockets\&. If
\m[blue]\fBbind interfaces only\fR\m[]
is set then
nmbd
will check the source address of any packets coming in on the broadcast sockets and discard any that don\*(Aqt match the broadcast addresses of the interfaces in the
\m[blue]\fBinterfaces\fR\m[]
parameter list\&. As unicast packets are received on the other sockets it allows
nmbd
to refuse to serve names to machines that send packets that arrive through any interfaces not listed in the

i586/porteus/base/003-mate/usr/share/system-tools-backends-2.0/scripts/Shares/SMB.pm
#-*- Mode: perl; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*-
#
# Copyright (C) 2000-2001 Ximian, Inc.
#
# Authors: Hans Petter Jansson <hpj@ximian.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU Library General Public License as published
# by the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Library General Public License for more details.
#
# You should have received a copy of the GNU Library General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.

package Shares::SMB;

use Utils::File;

# --- share_export_smb_info; information on a particular SMB export --- #

sub gst_share_smb_info_set
{
  my ($info, $key, $value) = @_;
  
  if ($value eq "")
  {
    delete $info->{$key};
  }
  else
  {
    $info->{$key} = $value;
  }
}

sub gst_share_smb_info_get_name
{
  return $_[0]->{'name'};
}

sub gst_share_smb_info_set_name
{
  &gst_share_smb_info_set ($_[0], 'name', $_[1]);
}

sub gst_share_smb_info_get_point
{
  return $_[0]->{'point'};
}

sub gst_share_smb_info_set_point
{
  &gst_share_smb_info_set ($_[0], 'point', $_[1]);
}

sub gst_share_smb_info_get_comment
{
  return $_[0]->{'comment'};
}

sub gst_share_smb_info_set_comment
{
  &gst_share_smb_info_set ($_[0], 'comment', $_[1]);
}

sub gst_share_smb_info_get_enabled
{
  return $_[0]->{'enabled'};
}

sub gst_share_smb_info_set_enabled
{
  &gst_share_smb_info_set ($_[0], 'enabled', $_[1]);
}

sub gst_share_smb_info_get_browse
{
  return $_[0]->{'browse'};
}

sub gst_share_smb_info_set_browse
{
  &gst_share_smb_info_set ($_[0], 'browse', $_[1]);
}

sub gst_share_smb_info_get_public
{
  return $_[0]->{'public'};
}

sub gst_share_smb_info_set_public
{
  &gst_share_smb_info_set ($_[0], 'public', $_[1]);
}

sub gst_share_smb_info_get_write
{
  return $_[0]->{'write'};
}

sub gst_share_smb_info_set_write
{
  &gst_share_smb_info_set ($_[0], 'write', $_[1]);
}


# --- share_smb_table; multiple instances of share_smb_info --- #

sub smb_table_find
{
  my ($name, $shares) = @_;

  foreach $i (@$shares)
  {
    return $i if ($$i[0] eq $name)
  }

  return undef;
}

sub get_distro_smb_file
{
  my ($smb_comb);

  my %dist_map =
  (
   "debian"          => "debian",
   "redhat-6.2"      => "redhat-6.2",
   "redhat-7.0"      => "debian",
   "redhat-7.1"      => "debian",
   "redhat-7.2"      => "debian",
   "redhat-7.3"      => "debian",
   "redhat-8.0"      => "debian",
   "mandrake-9.0"    => "debian",
   "suse-9.0"        => "debian",
   "slackware-9.1.0" => "debian",
   "slackware-14.0"  => "debian",
   "slackware-14.1"  => "debian",
   "gentoo"          => "debian",
   "archlinux"       => "debian",
   "pld-1.0"         => "pld-1.0",
   "vine-3.0"        => "debian",
   "freebsd-5"       => "freebsd-5",
  );

  my %dist_tables =
  (
   "redhat-6.2" => "/etc/smb.conf",
   "debian" => "/etc/samba/smb.conf",
   "pld-1.0"    => "/etc/smb/smb.conf",
   "freebsd-5"  => "/usr/local/etc/smb.conf",
  );

  my $dist = $dist_map {$Utils::Backend::tool{"platform"}};
  return $dist_tables{$dist} if $dist;
  return undef;
}

sub get_share_info
{
  my ($smb_conf_name, $section) = @_;
  my @share;

  push @share, $section;
  push @share, &Utils::Parse::get_from_ini      ($smb_conf_name, $section, "path");
  push @share, &Utils::Parse::get_from_ini      ($smb_conf_name, $section, "comment");
  push @share, &Utils::Parse::get_from_ini_bool ($smb_conf_name, $section, "available");
  push @share, &Utils::Parse::get_from_ini_bool ($smb_conf_name, $section, "browsable") ||
               &Utils::Parse::get_from_ini_bool ($smb_conf_name, $section, "browseable");
  push @share, &Utils::Parse::get_from_ini_bool ($smb_conf_name, $section, "public")      ||
               &Utils::Parse::get_from_ini_bool ($smb_conf_name, $section, "guest");
  push @share, &Utils::Parse::get_from_ini_bool ($smb_conf_name, $section, "writable")    ||
               &Utils::Parse::get_from_ini_bool ($smb_conf_name, $section, "writeable");

  return \@share;
}

sub set_share_info
{
  my ($smb_conf_file, $share) = @_;
  my ($section);

  $section = shift (@$share);

  &Utils::Replace::set_ini        ($smb_conf_file, $section, "path",      shift (@$share));
  &Utils::Replace::set_ini        ($smb_conf_file, $section, "comment",   shift (@$share));
  &Utils::Replace::set_ini_bool   ($smb_conf_file, $section, "available", shift (@$share));
  &Utils::Replace::set_ini_bool   ($smb_conf_file, $section, "browsable", shift (@$share));
  &Utils::Replace::set_ini_bool   ($smb_conf_file, $section, "public",    shift (@$share));
  &Utils::Replace::set_ini_bool   ($smb_conf_file, $section, "writable",  shift (@$share));

  &Utils::Replace::remove_ini_var ($smb_conf_file, $section, "browseable");
  &Utils::Replace::remove_ini_var ($smb_conf_file, $section, "guest");
  &Utils::Replace::remove_ini_var ($smb_conf_file, $section, "writeable");
}

sub get_shares
{
  my ($smb_conf_file) = @_;
  my (@sections, @table, $share);

  # Get the sections.
  @sections = &Utils::Parse::get_ini_sections ($smb_conf_file);

  for $section (@sections)
  {
    next if ($section =~ /^(global)|(homes)|(printers)|(print\$)$/);
    next if (&Utils::Parse::get_from_ini_bool ($smb_conf_file, $section, "printable"));

    $share = &get_share_info ($smb_conf_file, $section);
    push @table, $share;
  }

  return \@table;
}

sub get_smb_users
{
  my ($fd, @info, $users);

  $fd = &Utils::File::run_pipe_read ("pdbedit -L");
  return [] if (!$fd);

  while (<$fd>)
  {
    chomp;
    @info = split (/:/, $_);
    push @$users, [ $info[0], "" ];
  }

  return $users;
}

sub get
{
  my ($smb_conf_file);
  my ($shares, $workgroup, $desc, $wins, $winsserver, $users);

  $smb_conf_file = &get_distro_smb_file;
  $shares = &get_shares ($smb_conf_file);

  $workgroup = &Utils::Parse::get_from_ini ($smb_conf_file, "global", "workgroup");
  $smbdesc = &Utils::Parse::get_from_ini ($smb_conf_file, "global", "server string");
  $wins = &Utils::Parse::get_from_ini_bool ($smb_conf_file, "global", "wins support");
  $winsserver = &Utils::Parse::get_from_ini ($smb_conf_file, "global", "wins server");
  $users = &get_smb_users ();

  return ($shares, $workgroup, $smbdesc, $wins, $winsserver, $users);
}

sub set_shares
{
  my ($smb_conf_file, $shares) = @_;
  my (@sections, $section, $share);

  # Get the sections.
  @sections = &Utils::Parse::get_ini_sections ($smb_conf_file);

  # remove deleted sections
  foreach $section (@sections)
  {
    next if ($section =~ /^(global)|(homes)|(printers)|(print\$)$/);
    next if (&Utils::Parse::get_from_ini_bool ($smb_conf_file, $section, "printable"));

    if (!&smb_table_find ($section, $shares))
    {
      Utils::Replace::remove_ini_section ($smb_conf_file, $section);
    }
  }

  for $share (@$shares)
  {
    &set_share_info ($smb_conf_file, $share);
  }
}

sub set_smb_users
{
  my ($users) = @_;
  my ($old_users, $user, $config, $state);
  my ($hash, $old_hash, $pipe);

  $old_users = &get_smb_users ();

  foreach $user (@$users)
  {
    $$config{$$user[0]} |= 1;
    $hash{$$user[0]} = $user;
  }

  foreach $user (@$old_users)
  {
    $$config{$$user[0]} |= 2;
    $old_hash{$$user[0]} = $user;
  }

  foreach $i (keys %$config)
  {
    $state = $$config{$i};
    $user = $hash{$i};

    if ($state == 1 || ($state == 3 && $$user[1]))
    {
      # User added, or password was modified
      $user = $hash{$i};
      $pipe = &Utils::File::run_pipe_write ("smbpasswd -s -a $$user[0]");
      # Have to write the password twice
      print $pipe "$$user[1]\n";
      print $pipe "$$user[1]\n";
      &Utils::File::close_file ($pipe);
    }
    elsif ($state == 2)
    {
      # User deleted
      $user = $old_hash{$i};
      &Utils::File::run ("pdbedit", "--delete", "-u", $$user[0]);
    }
  }
}

sub set
{
  my ($shares, $workgroup, $desc, $wins, $winsserver, $users) = @_;
  my ($smb_conf_file);
  my (@sections, $export);

  $smb_conf_file = &get_distro_smb_file;

  &set_shares ($smb_conf_file, $shares);

  &Utils::Replace::set_ini ($smb_conf_file, "global", "workgroup", $workgroup);
  &Utils::Replace::set_ini ($smb_conf_file, "global", "server string", $desc);
  &Utils::Replace::set_ini_bool ($smb_conf_file, "global", "wins support", $wins);
  &Utils::Replace::set_ini ($smb_conf_file, "global", "wins server", ($wins) ? "" : $winsserver);

  &set_smb_users ($users);
}

sub get_files
{
  my ($files);

  push @$files, &get_distro_smb_file ();
  return $files;
}

1;
